README.md
# create-image

[![Tests](https://img.shields.io/github/actions/workflow/status/LinqLover/create-image/test.yml?label=%F0%9F%A7%AA%20Test)](https://github.com/LinqLover/create-image/actions/workflows/test.yml)
[![Lint](https://img.shields.io/github/actions/workflow/status/LinqLover/create-image/lint.yml?label=%F0%9F%A7%B5%20Lint)](https://github.com/LinqLover/create-image/actions/workflows/lint.yml)
[![Public workflows that use this action](https://img.shields.io/endpoint?url=https%3A%2F%2Fused-by.vercel.app%2Fapi%2Fgithub-actions%2Fused-by%3Faction%3DLinqLover%2Fcreate-image%26badge%3Dtrue)](https://sourcegraph.com/search?q=context:global+LinqLover/create-image+file:.github/workflows&patternType=literal)

> GitHub Action for creating and preparing an all-in-one image for [Squeak](https://squeak.org/).

You can use this action in your workflow to automatically deploy a **one-click/all-in-one image bundle of Squeak** containing **your app or modifications.**
- The bundle is fetched from <https://squeak.org/downloads> and will contain a **ready-to-use image** and **VM binaries** for all supported platforms (head over to [squeak-smalltalk/squeak-app](https://github.com/squeak-smalltalk/squeak-app) for more information).
- You can run **custom scripts** to **prepare** the image before saving it or to **postpare** it when it is opened again.
- You can choose the **Squeak version** and **bitness** of the image.

## Usage

### Example workflow

Below is a simple example workflow that creates an image, loads some code into it, and uploads the image bundle as an artifact of your workflow:

```yml
on:
  push:
    branches: [master]

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: LinqLover/create-image@latest
        id: create-image
        with:
          squeak-version: 6.0
          prepare-script: ./scripts/prepareImage.st
      - uses: actions/upload-artifact@master
        with:
          name: image
          path: ${{ steps.create-image.outputs.bundle-path }}
```

In this example, `scripts/prepareImage.st` might look like this:

```smalltalk
"Install something..."
Installer new merge: #ffi.

"Open some welcome contents..."
MovingEyeMorph extraExampleSqueakIsWatchingYou openCenteredInWorld.
```

### Inputs

<table>
  <thead>
    <tr>
      <td>Parameter</td>
      <td>Description</td>
      <td>Example</td>
      <td>Required?</td>
    </tr>
  <tbody>
    <tr>
      <td><code>squeak-version</code></td>
      <td>The version of Squeak to be used.</td>
      <td><code>trunk</code>, <code>'6.0'</code></td>
      <td><i>required</i></td>
    </tr>
    <tr>
      <td><code>squeak-bitness</code></td>
      <td>The bitness of the image to be created. Defaults to <code>64</code>.</td>
      <td><code>64</code>, <code>32</code> (⚠ currently not supported on GitHub Actions, see <a href="https://github.com/actions/runner/issues/1181">actions/runner#1181</a>)</td>
      <td><i>optional</i></td>
    </tr>
    <tr>
      <td><code>prepare-script</code></td>
      <td>A script to be filed into the image before saving it.</td>
      <td><code>/path/to/script.st</code></td>
      <td><i>optional</i></td>
    </tr>
    <tr>
      <td><code>postpare-script</code></td>
      <td>A script to be executed in the image after saving it, i.e., at the moment the user opens it again.</td>
      <td><code>/path/to/script.st</code></td>
      <td><i>optional</i></td>
    </tr>
  </tbody>
</table>

All the silly mistakes and typos made by you in the prepare/postpare scripts will be caught by the action and displayed in the action log.

### Outputs

<table>
  <thead>
    <tr>
      <td>Parameter</td>
      <td>Description</td>
      <td>Example</td>
    </tr>
  <tbody>
    <tr>
      <td><code>bundle-path</code></td>
      <td>Indicates the path to the created bundle file.</td>
      <td><code>/path/to/Squeak6.0Alpha-12345-64bit-AllInOne.zip</code></td>
    </tr>
  </tbody>
</table>

To browse further usage examples, click [here](https://sourcegraph.com/search?q=context:global+LinqLover/create-image+file:.github/workflows&patternType=literal).

## Contribution

... is as always very welcome! If you use this action, have any ideas for improvements, or even would like to submit your patch, my issues & PRs are open!

---

Carpe Squeak! 🎈


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v1.1]

### Fixed

- Fix default FileDirectory in which the prepare script will run

## [v1.2]

### Fixed

- Align action to breaking changes in the `./squeak.sh` invocation script

## [v2.0]

### Added

- Add support for configuring Squeak version and bitness
- Reduce noise on action log


test/extract.sh
#!/usr/bin/env bash
set -eo pipefail
output="output"

error() {
    >&2 echo "$@"
    exit 1
}

test -n "$1" || error "invalid archive file"

extracted="$output/image"
mkdir -p "$extracted"
unzip -q "$1" -d "$extracted"

ls "$extracted"
test -f "$extracted/squeak.bat"
test -f "$extracted/squeak.sh"

echo "extracted image to $extracted"


test/test_local.sh
#!/usr/bin/env bash
# Rough local equivalent of .github/workflows/test.yml
# 🤞 Crossing fingers that this script will not become out of sync with the workflow file
set -eo pipefail

if [[ -z "$SQUEAK_VERSION" ]]; then
	versions=(trunk 6.0 5.3)
else
	versions=("$SQUEAK_VERSION")
fi
if [[ -z "$SQUEAK_BITNESS" ]]; then
	bitnesses=(64)
else
	bitnesses=("$SQUEAK_BITNESS")
fi

for version in "${versions[@]}"; do
	for bitness in "${bitnesses[@]}"; do
		echo "Testing Squeak $version $bitness"
		output=$(SQUEAK_VERSION="$version" SQUEAK_BITNESS="$bitness" PREPARE_SCRIPT=test/prepare.st POSTPARE_SCRIPT=test/postpare.st GITHUB_OUTPUT=/dev/stdout src/create_image.sh | tee /dev/tty)
		bundlePath=$(echo "$output" | sed -n 's/.*bundle-path=\(.*\)/\1/p')
		test/extract.sh "$bundlePath"
		output/image/squeak.sh "$(realpath test/test.st)" -- "$version" "$bitness"
		rm -rf output/image
	done
done


src/create_image.sh
#!/usr/bin/env bash
set -eo pipefail

# Read & validate inputs
if [[ -z "$SQUEAK_VERSION" ]]; then
    echo "SQUEAK_VERSION is not set"
    exit 1
fi
squeakVersion="$SQUEAK_VERSION"
squeakBitness=64
if [[ "$SQUEAK_BITNESS" ]]; then
    if [[ "$SQUEAK_BITNESS" != 32 && "$SQUEAK_BITNESS" != 64 ]]; then
        echo "SQUEAK_BITNESS must be 32 or 64"
        exit 1
    fi
    squeakBitness="$SQUEAK_BITNESS"
fi
if [[ "$squeakBitness" == 32 ]]; then
    echo "WARNING: 32bit may not be supported on GitHub Actions"
fi
if [[ "$PREPARE_SCRIPT" ]]; then
    prepareScript="$(realpath "$PREPARE_SCRIPT")"
fi
if [[ "$POSTPARE_SCRIPT" ]]; then
    postpareScript="$(realpath "$POSTPARE_SCRIPT")"
fi

script_dir="$(realpath "$(dirname "$0")")"
cd "$script_dir"
fileinScript="$(realpath "prepareImage.st")"
mkdir -p ../output && cd ../output

# Download and extract latest Trunk release
wgetArgs=(--progress=dot:giga)
files_server="http://files.squeak.org/$squeakVersion"
files_index="files.html"
wget "${wgetArgs[@]}" -O "$files_index" "$files_server"
build="$(grep -P -o "(?<=a href=\")Squeak[^<>]*?-${squeakBitness}bit(?=\/\")" "$files_index" | tail -1)"
buildAio="$build-All-in-One.zip"
dir="$build-All-in-One/"
wget "${wgetArgs[@]}" "$files_server/$build/$buildAio"
unzip -q \
    -d "$dir" "$buildAio" \
    -x '*.mo'  # skip superfluous localization files (optimization)
echo "bundle-path=$(realpath "$buildAio")" >> "$GITHUB_OUTPUT"

# Prepare image
cd "$dir"
if [[ "$CI" == true ]]; then
    VMOPTIONS="-headless"
fi
./squeak.sh $VMOPTIONS "$fileinScript" "$prepareScript" "$postpareScript"

# Clean up caches
shopt -s globstar
rm -rf ./**/{{package,github}-cache/,\#tmp\#*}

# Write changes back to zip
zip \
    -u -r \
    "../$buildAio" .
# TODO optimize: Ignore *.mo files here again (-x does not work)

# Remove temporary files
cd ..
rm -rf "$dir"
